File: smoke_tests/csharp/path_traversal.cs
// Path Traversal vulnerabilities in C#
using System;
using System.IO;

public class PathTraversalVulnerabilities
{
    public string ReadFileUnsafe(string filename)
    {
        // VULNERABLE: Path traversal via concatenation
        string path = "/var/data/" + filename;
        return File.ReadAllText(path);
    }

    public byte[] DownloadFileUnsafe(string fileName)
    {
        // VULNERABLE: No path validation
        string basePath = @"C:\uploads\";
        string fullPath = Path.Combine(basePath, fileName);
        return File.ReadAllBytes(fullPath);
    }

    public void DeleteFileUnsafe(string userPath)
    {
        // VULNERABLE: Arbitrary file deletion
        string path = Path.Combine("/tmp/", userPath);
        File.Delete(path);
    }

    public void SaveFileUnsafe(string filename, byte[] data)
    {
        // VULNERABLE: User controls save location
        string path = @"C:\uploads\" + filename;
        File.WriteAllBytes(path, data);
    }

    public FileStream OpenFileUnsafe(string requestedFile)
    {
        // VULNERABLE: Direct path usage
        return new FileStream("/public/files/" + requestedFile, FileMode.Open);
    }
}

File: smoke_tests/rust/path_traversal.rs
// Path Traversal vulnerability in Rust
use std::fs;
use std::path::Path;

fn read_file_unsafe(filename: &str) -> Result<String, std::io::Error> {
    // VULNERABLE: No path validation
    let path = format!("/var/data/{}", filename);
    fs::read_to_string(path)
}

fn serve_file_unsafe(user_path: &str) -> Result<Vec<u8>, std::io::Error> {
    // VULNERABLE: Direct path concatenation
    let base = "/public/files/";
    let full_path = format!("{}{}", base, user_path);
    fs::read(full_path)
}

fn delete_file_unsafe(filename: &str) -> Result<(), std::io::Error> {
    // VULNERABLE: Arbitrary file deletion
    let path = Path::new("/uploads").join(filename);
    fs::remove_file(path)
}
